### DISCUSSION TITLE
SB5.1 - OCI - Ansible variables for Deploy/managed-learn/ml-* jobs

### DISCUSSION THREAD

**Discussion by ddevadat** [Sunbird-Ed/Community#394] [Installation]
We are deploying sunbird on OCI at release 5.1

For the below jobs we need the ansible variables values
Deploy/managed-learn/ml-analytics-service
Deploy/managed-learn/ml-core-service
Deploy/managed-learn/ml-projects-service

1) Deploy/managed-learn/ml-analytics-service

When we run this job, what should be the value of   ml_analytics_version ?

2) Deploy/managed-learn/ml-core-service
We need to provide the bucket name for the dictionary ml_cloud_config. what should be the value

3) Deploy/managed-learn/ml-projects-service 
For this job we are seeing the below errors in the pods

```
│ APPLICATION_PORT            │ PASSED │
├─────────────────────────────┼────────┤
│ APPLICATION_ENV             │ PASSED │
├─────────────────────────────┼────────┤
│ MONGODB_URL                 │ PASSED │
├─────────────────────────────┼────────┤
│ INTERNAL_ACCESS_TOKEN       │ PASSED │
├─────────────────────────────┼────────┤
│ KAFKA_COMMUNICATIONS_ON_OFF │ PASSED │
├─────────────────────────────┼────────┤
│ KAFKA_URL                   │ PASSED │
├─────────────────────────────┼────────┤
│ USER_SERVICE_URL            │ PASSED │
├─────────────────────────────┼────────┤
│ SERVICE_NAME                │ PASSED │
├─────────────────────────────┼────────┤
│ CERTIFICATE_SERVICE_URL     │ PASSED │
├─────────────────────────────┼────────┤
│ PROJECT_CERTIFICATE_ON_OFF  │ PASSED │
└─────────────────────────────┴────────┘
issuer Kid url :  http://registry-service:8081/api/v1/PublicKey/search
issuer Kid bodyData :  {"filters":{}}
Environment : dev
Application is running on the port : 3000
Connected to DB
Connected to Kafka
KID rc call error :  connect ECONNREFUSED 10.233.7.135:8081
Server stoped . Failed to set certificate issuer Kid value
```

**Comment by ddevadat** [comment]
any update on this

**Reply by gandham-santhosh** [reply]
@aks30 Can you please check and respond to this?

cc: @vijiurs @chitranshu-keshav

**Comment by aks30** [comment]
For Cloud storage configuration in ml-core-service in OCI please refer to the section "# Oracle Cloud Configuration   " at the end -https://github.com/project-sunbird/ml-core-service/blob/release-5.1.0/.env.sample

For ML Project service errors do check and confirm if Sunbird RC deployment is done as per release notes in 5.1 here - https://ed.sunbird.org/use/updating-sunbird-releases/5.0.0-to-5.1.0#sunbird-rc Also in case it is not done to skip the same during the setup we can set the key PROJECT_CERTIFICATE_ON_OFF to be "OFF". Refer- https://github.com/project-sunbird/ml-projects-service/blob/release-5.1.0/.env.sample

**Reply by ddevadat** [reply]
> For Cloud storage configuration in ml-core-service in OCI please refer to the section "# Oracle Cloud Configuration " at the end -https://github.com/project-sunbird/ml-core-service/blob/release-5.1.0/.env.sample
> 
> For ML Project service errors do check and confirm if Sunbird RC deployment is done as per release notes in 5.1 here - https://ed.sunbird.org/use/updating-sunbird-releases/5.0.0-to-5.1.0#sunbird-rc Also in case it is not done to skip the same during the setup we can set the key PROJECT_CERTIFICATE_ON_OFF to be "OFF". Refer- https://github.com/project-sunbird/ml-projects-service/blob/release-5.1.0/.env.sample


Which bucket is assigned for ml-core-service? Do we need to use an existing bucket ? If not , Does it have to be a public bucket?

**Reply by aks30** [reply]
@ddevadat the bucket to use will be as different per instance of Sunbird, you can have it as per your needs.

**Reply by ddevadat** [reply]
> @ddevadat the bucket to use will be as different per instance of Sunbird, you can have it as per your needs.

This is for Sunbird-Ed. Can we give any random name for the bucket or is there a standard bucket name? Also is it a private or public bucket.

**Reply by aks30** [reply]
I suggest use a name which makes sense, in this case can be {{instance_name}}-manage-learn and it will be a public bucket

**Comment by aks30** [comment]
@Shakthieshwari Can you check for the issue mentioned for ml-analytics service once?

**Reply by Shakthieshwari** [reply]
@ddevadat value of ml_analytics_version should be **release-5.1.0_RC14**.

Please refer this link :- https://ed.sunbird.org/use/updating-sunbird-releases/5.0.0-to-5.1.0#5.0.0-to-5.1.0

**Reply by ddevadat** [reply]
> @ddevadat value of ml_analytics_version should be **release-5.1.0_RC14**.
> 
> Please refer this link :- https://ed.sunbird.org/use/updating-sunbird-releases/5.0.0-to-5.1.0#5.0.0-to-5.1.0

What vm should be specified in the inventory host for managed-learn for the group
[ml-analytics-service]

**Reply by Krishnaj20** [reply]
@aks30 @Shakthieshwari - Please respond to this query.

**Reply by Shakthieshwari** [reply]
@ddevadat @Krishnaj20 I didnot understand the Query, Can you please elaborate ?

**Reply by ddevadat** [reply]
> @ddevadat @Krishnaj20 I didnot understand the Query, Can you please elaborate ?

When i executed the job..it failed with the error

```
PLAY [ml-analytics-service] ****************************************************

TASK [Gathering Facts] *********************************************************
fatal: [10.0.0.45]: UNREACHABLE! => {
    "changed": false,
    "unreachable": true
}

MSG:

Failed to connect to the host via ssh: ssh: connect to host 10.0.0.45 port 22: Connection timed out
```

This is because in the inventory host ...there is a group defined like
```
[ml-analytics-service]
10.0.0.45
```

10.0.045 -- seems to be a dummy value.
My question is which vm should be assigned to the group ml-analytics-service?

**Reply by raghupathiguduri** [reply]
@ddevadat  can you please add the VM Ip in which you want to deploy the ml-analytics-service. You can have a seperate VM for ml-analytics-service.

cc: @Shakthieshwari

**Comment by ddevadat** [comment]
Thanks for all the response.. I am putting together the issue status so far.
1) Deploy/managed-learn/ml-analytics-service 

Add the host details in the group [ml-analytics-service] and ml_analytics_version = release-5.1.0_RC14.

The job still failed with the below error


```
TASK [ml-analytics-service : CHANGE THE OWNERSHIP FOR THIS /opt/sparkjobs DIRECTORY] ***
changed: [10.0.1.9] => (item=/opt/sparkjobs)
ok: [10.0.1.9] => (item=/opt/sparkjobs)
changed: [10.0.1.9] => (item=/opt/sparkjobs/faust_as_service)

TASK [ml-analytics-service : Delete the virtualenv DIR] ************************
[WARNING]: Consider using the file module with state=absent rather than running
'rm'.  If you need to use command because file is insufficient you can add
'warn: false' to this command task or set 'command_warnings=False' in
ansible.cfg to get rid of this message.

changed: [10.0.1.9]

TASK [ml-analytics-service : Install python virtual environment] ***************
fatal: [10.0.1.9]: FAILED! => {
    "changed": true,
    "cmd": "cd /opt/sparkjobs && virtualenv --python=python3.8 spark_venv",
    "delta": "0:00:00.001430",
    "end": "2023-02-16 23:49:01.027135",
    "rc": 127,
    "start": "2023-02-16 23:49:01.025705"
}

STDERR:

/bin/sh: 1: virtualenv: not found


MSG:

non-zero return code
```


2) Deploy/managed-learn/ml-core-service

Update the bucket details for the dictionary ml_cloud_config

The job got successful, but there are errors in the pods logs like below. Its still referring to an azure based url . Where is this configured
```

├──────────────────────────┼────────┤
│ OCI_SECRET_ACCESS_KEY    │ PASSED │
├──────────────────────────┼────────┤
│ OCI_BUCKET_NAME          │ PASSED │
├──────────────────────────┼────────┤
│ OCI_BUCKET_REGION        │ PASSED │
├──────────────────────────┼────────┤
│ OCI_BUCKET_ENDPOINT      │ PASSED │
└──────────────────────────┴────────┘
Environment: dev
Application is running on the port:3000
Connected to database!
(node:7) UnhandledPromiseRejectionWarning: RestError: request to https://.blob.core.windows.net/sl-dev-storage?restype=container failed, reason: getaddrinfo ENOTFOUND .blob.core.windows.net
    at NodeFetchHttpClient.sendRequest (/opt/core/node_modules/@azure/core-http/dist/index.js:2684:23)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:7) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:7) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

3) Deploy/managed-learn/ml-projects-service

I verified that Sunbird RC deployment is done , but still getting the errors
issuer Kid url :  http://registry-service:8081/api/v1/PublicKey/search
issuer Kid bodyData :  {"filters":{}}
Environment : dev
Application is running on the port : 3000
Connected to DB
Connected to Kafka
KID rc call error :  connect ECONNREFUSED 10.233.7.135:8081
Server stoped . Failed to set certificate issuer Kid value

**Reply by Krishnaj20** [reply]
@aks30 - can you respond to this query?

@vijiurs @gandham-santhosh @chitranshu-keshav

**Reply by Shakthieshwari** [reply]
@raghupathiguduri Please respond to the query asked on the ML Analytics Service

**Reply by aks30** [reply]
@ddevadat For ml core service can you share the complete startup logs, not able to gain a complete idea what is happening, possibly add screenshots if not possible to mention all in the text here.

For the Sunbird RC setup - It seems the project service is not able to communicate with the RC service. Can you confirm if the URL for any external service to connect with RC service is correct - "http://registry-service:8081" basically from any other service if RC Service APIs have to be called what would be the URL

**Reply by Krishnaj20** [reply]
tagging @gandham-santhosh to this thread. 
Santhosh - pls check this query.

**Comment by ddevadat** [comment]
The logs from ml-core-service

```

jenkins@jenkins:~$ k logs ml-core-service-b9c594dcf-pxjzr
[Thu Feb 16 2023 23:54:06 GMT+0000 (Coordinated Universal Time)] INFO ┌──────────────────────────┬────────┐
│ APPLICATION_PORT         │ PASSED │
├──────────────────────────┼────────┤
│ APPLICATION_ENV          │ PASSED │
├──────────────────────────┼────────┤
│ MONGODB_URL              │ PASSED │
├──────────────────────────┼────────┤
│ INTERNAL_ACCESS_TOKEN    │ PASSED │
├──────────────────────────┼────────┤
│ CLOUD_STORAGE            │ PASSED │
├──────────────────────────┼────────┤
│ GCP_PATH                 │ PASSED │
├──────────────────────────┼────────┤
│ GCP_BUCKET_NAME          │ PASSED │
├──────────────────────────┼────────┤
│ AZURE_ACCOUNT_NAME       │ PASSED │
├──────────────────────────┼────────┤
│ AZURE_ACCOUNT_KEY        │ PASSED │
├──────────────────────────┼────────┤
│ AZURE_STORAGE_CONTAINER  │ PASSED │
├──────────────────────────┼────────┤
│ AWS_ACCESS_KEY_ID        │ PASSED │
├──────────────────────────┼────────┤
│ AWS_SECRET_ACCESS_KEY    │ PASSED │
├──────────────────────────┼────────┤
│ AWS_BUCKET_NAME          │ PASSED │
├──────────────────────────┼────────┤
│ AWS_BUCKET_REGION        │ PASSED │
├──────────────────────────┼────────┤
│ AWS_BUCKET_ENDPOINT      │ PASSED │
├──────────────────────────┼────────┤
│ KEYCLOAK_PUBLIC_KEY_PATH │ PASSED │
├──────────────────────────┼────────┤
│ ML_SURVEY_SERVICE_URL    │ PASSED │
├──────────────────────────┼────────┤
│ ML_PROJECT_SERVICE_URL   │ PASSED │
├──────────────────────────┼────────┤
│ USER_SERVICE_URL         │ PASSED │
├──────────────────────────┼────────┤
│ CSV_REPORTS_PATH         │ PASSED │
├──────────────────────────┼────────┤
│ APP_PORTAL_BASE_URL      │ PASSED │
├──────────────────────────┼────────┤
│ FORM_SERVICE_URL         │ PASSED │
├──────────────────────────┼────────┤
│ OCI_ACCESS_KEY_ID        │ PASSED │
├──────────────────────────┼────────┤
│ OCI_SECRET_ACCESS_KEY    │ PASSED │
├──────────────────────────┼────────┤
│ OCI_BUCKET_NAME          │ PASSED │
├──────────────────────────┼────────┤
│ OCI_BUCKET_REGION        │ PASSED │
├──────────────────────────┼────────┤
│ OCI_BUCKET_ENDPOINT      │ PASSED │
└──────────────────────────┴────────┘
Environment: dev
Application is running on the port:3000
Connected to database!
(node:7) UnhandledPromiseRejectionWarning: RestError: request to https://.blob.core.windows.net/sl-dev-storage?restype=container failed, reason: getaddrinfo ENOTFOUND .blob.core.windows.net
    at NodeFetchHttpClient.sendRequest (/opt/core/node_modules/@azure/core-http/dist/index.js:2684:23)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
(node:7) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:7) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code
```.

**Reply by chitranshu-keshav** [reply]
@aks30 Can you please get back on the above

**Reply by aks30** [reply]
@ddevadat Can you further share the config map or the env set for this service while deployment? CLOUD_STORAGE value for you I am guessing should be "OCI" , maybe its "AZURE" for now

**Comment by ddevadat** [comment]
This is the value in the configmap

```
jenkins@jenkins:~$ k get cm ml-core-service-config -o yaml
apiVersion: v1
data:
  APP_PORTAL_BASE_URL: https://XXX.XXX.XXX.XXX
  APPLICATION_ENV: dev
  APPLICATION_PORT: "3000"
  CLOUD_STORAGE: OCI
  ELASTICSEARCH_COMMUNICATIONS_ON_OFF: "OFF"
  ELASTICSEARCH_ENTITIES_INDEX: ""
  ELASTICSEARCH_HOST_URL: ""
  ENABLE_CONSOLE_LOGGING: "OFF"
  GOTENBERG_SERVICE_URL: http://gotenberg-service:3000
  INTERNAL_ACCESS_TOKEN: XXXX
  KEYCLOAK_PUBLIC_KEY_PATH: keycloak-public-keys
  ML_PROJECT_SERVICE_URL: http://ml-projects-service:3000
  ML_SURVEY_SERVICE_URL: http://ml-survey-service:3000
  MONGODB_URL: mongodb://XXXXX:27017/ml-survey
  OCI_ACCESS_KEY_ID: XXXXXX
  OCI_BUCKET_ENDPOINT: https://XXXXX.compat.objectstorage.ap-hyderabad-1.oraclecloud.com
  OCI_BUCKET_NAME: sunbird-prod-XXXXX-manage-learn
  OCI_BUCKET_REGION: ap-hyderabad-1
  OCI_SECRET_ACCESS_KEY: XXXXX
  USER_SERVICE_URL: http://learner-service:9000
```

**Reply by gandham-santhosh** [reply]
Somewhere is it trying to load azure storage url base url ".blob.core.windows.net" 
@aks30 is there any place this is hardcoded?

**Reply by VISHNUDAS-tunerlabs** [reply]
@ddevadat It seems to be a warning, it will not be an issue as in the service would work fine, the warning we will try and address in next release.

**Comment by ddevadat** [comment]
@aks30 
For the ml-project-service, still getting the below error
```

│ issuer Kid url :  http://registry-service:8081/api/v1/PublicKey/search                                                                                                                                         │
│ issuer Kid bodyData :  {"filters":{}}                                                                                                                                                                          │
│ Environment : dev                                                                                                                                                                                              │
│ Application is running on the port : 3000                                                                                                                                                                      │
│ Connected to DB                                                                                                                                                                                                │
│ Connected to Kafka                                                                                                                                                                                             │
│ KID rc call error :  connect ECONNREFUSED XX.XX.XX.XX:8081                                                                                                                                                    │
│ Server stoped . Failed to set certificate issuer Kid value
```

I tried to add the flag 
`PROJECT_CERTIFICATE_ON_OFF = "OFF"`

but it didnt help

I want to know how does it get certificate issuer kid
```
# certificate issuer KID value
CERTIFICATE_ISSUER_KID={{certificate_issuer_kid | default("")}}

```
i am not passing any variable named certificate_issuer_kid so its taking as ""

**Reply by aks30** [reply]
@ddevadat CERTIFICATE_ISSUER_KID  is something the service picks up automatically from Sunbird RC service at the time of starting. You can check on 3 things - 1) Confirm Sunbird RC is deployed as per https://ed.sunbird.org/use/updating-sunbird-releases/5.0.0-to-5.1.0#sunbird-rc its important to run build jobs first and then deploy as per comments. 2) Onboard relevant APIs - https://ed.sunbird.org/use/updating-sunbird-releases/5.0.0-to-5.1.0#add-below-apis-in-internal-kong-configuration 3) Communication between project service and registry service pods is working.

**Comment by ddevadat** [comment]
@aks30 ..so it looks like registry service is running..but on checking registry logs.

```

	... 33 common frames omitted
Caused by: org.postgresql.util.PSQLException: FATAL: database "registry" does not exist
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2553)
	at org.postgresql.core.v3.QueryExecutorImpl.readStartupMessages(QueryExecutorImpl.java:2665)
	at org.postgresql.core.v3.QueryExecutorImpl.<init>(QueryExecutorImpl.java:147)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:273)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:51)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:223)
	at org.postgresql.Driver.makeConnection(Driver.java:465)
```

my question which jobs provisions the database named registry

**Comment by ddevadat** [comment]
i created the database name registry. Now i am not seeing above error..but still the ml-project-service is showing the same error.
Is it possible to have a meeting discussion on this

**Comment by ddevadat** [comment]
Now the error i am getting is slight different

```
│ issuer Kid url :  http://registry-service:8081/api/v1/PublicKey/search                                                                                                                                         │
│ issuer Kid bodyData :  {"filters":{}}                                                                                                                                                                          │
│ Environment : dev                                                                                                                                                                                              │
│ Application is running on the port : 3000                                                                                                                                                                      │
│ Connected to DB                                                                                                                                                                                                │
│ Connected to Kafka                                                                                                                                                                                             │
│ KID success response :  []                                                                                                                                                                                     │
│ Server stoped . Failed to set certificate issuer Kid value
```

**Comment by ddevadat** [comment]
In registry database i can see the below tables

```
registry=# \dt
                   List of relations
 Schema |            Name             | Type  |  Owner
--------+-----------------------------+-------+---------
 public | V_Issuer_GROUP              | table | sunbird
 public | V_ProjectCertificate_GROUP  | table | sunbird
 public | V_PublicKey_GROUP           | table | sunbird
 public | V_Recipient_GROUP           | table | sunbird
 public | V_Schema_GROUP              | table | sunbird
 public | V_Signatory_GROUP           | table | sunbird
 public | V_TrainingCertificate_GROUP | table | sunbird
 public | V_Training_GROUP            | table | sunbird
 public | V_graph_global_config       | table | sunbird
(9 rows)
```

**Reply by aks30** [reply]
@ddevadat Can you share the output of below curl from any running pod ml-core or ml-survey or anything?

```
curl --location --request POST 'http://registry-service:8081/api/v1/PublicKey/search' \
--header 'Content-Type: application/json' \
--data-raw '{"filters": {}}'
```

**Comment by ddevadat** [comment]
@aks30 
i didnt return anything

```
 #  curl --location --request POST 'http://registry-service:8081/api/v1/PublicKey/search' --header 'Content-Type: application/json' --data-raw '{"filters": {}}'
[]/
```

**Reply by aks30** [reply]
@ddevadat I suggest building and deploying the latest tags of all ML Service Repos once as per here  - https://ed.sunbird.org/use/updating-sunbird-releases/5.0.0-to-5.1.0#sunbird_ed and try once. Note for Project service use tag - **release-5.1.0_RC32**  Also, post that do set PROJECT_CERTIFICATE_ON_OFF = "OFF" and try, it should stop looking for registry service and thereby stop throwing this error.

**Comment by Krishnaj20** [comment]
@aks30 @ddevadat - I believe this thread needs to be resolved via a call. 

@anandvarada - Can you facilitate this discussion please?

**Reply by Krishnaj20** [reply]
@aks30 @ddevadat - Not sure if this thread is resolved or not - If this thread is resolved pls mark it as answered. 

@chitranshu-keshav

**Reply by aks30** [reply]
All services are working fine so I believe its resolved. Requesting the author of the discussion to @ddevadat to confirm and mark the appropriate comment as an answer

**Comment by imvikysharma** [comment]
Based on the information provided, it seems that there are some issues with the ml-projects-service deployment. The error message suggests that the connection to the registry service was refused, which may indicate that the service is not running or is not accessible from the pod.

To resolve this issue, you can try the following steps:

Check if the registry service is running and accessible from the pod. You can do this by running a ping or curl command from the pod to the service endpoint.

Check if the correct registry service endpoint URL is provided in the KID issuer configuration. You can verify this by checking the ml_cloud_config dictionary in the playbook for the ml-projects-service deployment.

If the above steps do not resolve the issue, check the logs of the registry service and the ml-projects-service pods for any error messages that may provide more information about the root cause of the issue.

Once you have identified and resolved the issue, you should be able to successfully deploy the ml-projects-service.

Meanwhile I would like to recommend a book on AWS
https://www.amazon.com/dp/B0BVJKH8PY?maas=maas_adg_AA06FAB7196A55FE3DBBEE7958CED70A_afap_abs&ref_=aa_maas&tag=maas
